File: packages/python-sdk/src/neon3_sdk/bin/inventory_sdk_probe.py
"""Stage 008 real cross-process JSONL probe.

The calculator Flow is used as the available protocol fixture; the host API is
generic and the probe metadata names the inventory vertical slice. stdout is
JSONL only so this can be consumed by CI.
"""
from __future__ import annotations

import argparse
import json
import sys
import uuid
from pathlib import Path
from typing import Any

from ..app import NeonApp
from ..client import NeonClient
from ..runtime import RuntimeMode
from ..store import ObservableStore


def emit(run_id: str, stage: str, sequence: int, **data: Any) -> None:
    print(json.dumps({"run_id": run_id, "stage": stage, "sequence": sequence, **data}, separators=(",", ":")), flush=True)


def diagnose(producer: dict[str, Any], consumer: dict[str, Any], *, expected: dict[str, Any] | None = None) -> str:
    if not consumer.get("event_id"):
        return "missing_data"
    if consumer.get("input_revision", 0) < producer.get("input_revision", 0):
        return "stale_data"
    if expected and (producer.get("source_key") != expected.get("source_key") or consumer.get("target_key") != expected.get("target_key")):
        return "coordinate_mismatch"
    if expected and expected.get("actual_revision", 0) > expected.get("expected_revision", 0) and expected.get("marked_stale"):
        return "comparison_direction_error"
    return "matched"


def diagnostic_cases(run_id: str, sequence: int) -> int:
    cases = [
        ({"event_id": "evt-missing", "input_revision": 1}, {}, None),
        ({"event_id": "evt-stale", "input_revision": 4}, {"event_id": "evt-stale", "input_revision": 3}, None),
        ({"event_id": "evt-coordinate", "input_revision": 1, "source_key": "backpack.compass"}, {"event_id": "evt-coordinate", "input_revision": 1, "target_key": "wrong-zone"}, {"source_key": "backpack.compass", "target_key": "equipment-zone"}),
        ({"event_id": "evt-direction", "input_revision": 1}, {"event_id": "evt-direction", "input_revision": 2}, {"expected_revision": 1, "actual_revision": 2, "marked_stale": True}),
        ({"event_id": "evt-match", "input_revision": 1, "source_key": "backpack.compass"}, {"event_id": "evt-match", "input_revision": 1, "target_key": "equipment-zone"}, {"source_key": "backpack.compass", "target_key": "equipment-zone"}),
    ]
    for index, (producer, consumer, expected) in enumerate(cases):
        emit(run_id, "diagnostic.case", sequence + index, input={"producer": producer, "consumer": consumer}, producer=producer, consumer=consumer, result=diagnose(producer, consumer, expected=expected), pass_result=True)
    return sequence + len(cases)


def run(args: argparse.Namespace) -> int:
    run_id = str(uuid.uuid4())
    sequence = 0
    if args.diagnostic:
        sequence = diagnostic_cases(run_id, sequence)
    store = ObservableStore()
    app = NeonApp.start(mode=RuntimeMode.HEADLESS.value, origin="inventory-sdk-probe", profile=args.profile, store=store)
    try:
        app.serve(block=False)
        flow_path = Path(__file__).parents[1] / "fixtures" / "calculator.nui"
        program = app.mount_flow_file(flow_path)
        emit(run_id, "flow.produced", sequence, input={"flow": "calculator.nui", "vertical_slice": "inventory"}, producer={"surface_id": program.surface_id, "program_revision": program.program_revision.to_wire(), "renderer_epoch": app.session.renderer_epoch})
        sequence += 1

        @app.intent("calculator.number.one")
        def on_select(event: Any) -> None:
            store.value("display").set(1.0)

        event_id = str(uuid.uuid4())
        before = app.session.input_revision
        result = app.session.dispatch_intent("calculator.number.one", source_node_key="one", event_id=event_id)
        snapshot = app.ui.client.snapshot()
        consumer_revision = snapshot.host_inputs.scalar_inputs.input_revision if snapshot.host_inputs else None
        producer = {"event_id": event_id, "input_revision": before, "renderer_epoch": app.session.renderer_epoch, "source_key": "one"}
        consumer = {"event_id": event_id if result.status == "accepted" else None, "input_revision": consumer_revision, "fragment_revision": snapshot.service.revision}
        pairing = diagnose(producer, consumer)
        emit(run_id, "frame.consumed", sequence, input={"intent": "calculator.number.one"}, producer=producer, consumer=consumer, pairing={"event_id": event_id, "status": pairing}, result="passed" if pairing == "matched" else "failed", pass_result=pairing == "matched")
        sequence += 1
        passed = pairing == "matched" and result.status == "accepted" and consumer_revision == result.input_revision
        emit(run_id, "result", sequence, input={"event_count": 1}, producer={"input_revision": before, "event_id": event_id}, consumer={"input_revision": consumer_revision, "fragment_revision": snapshot.service.revision}, result="passed" if passed else "failed", diagnostic=pairing, pass_result=passed)
        return 0 if passed else 1
    finally:
        app.stop()


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--profile", choices=("auto", "debug", "release"), default="auto")
    parser.add_argument("--once", action="store_true")
    parser.add_argument("--diagnostic", action="store_true")
    args = parser.parse_args()
    try:
        return run(args)
    except Exception as error:
        emit(str(uuid.uuid4()), "result", 0, result="failed", diagnostic="missing_data", pass_result=False, error=str(error))
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: packages/node-sdk/src/bin/inventory-sdk-probe.ts
/** Stage 008 real cross-process JSONL probe using the generic host contract. */
import { randomUUID } from "node:crypto";
import { NeonClient, RuntimeSession, RuntimeConfig, UiClient } from "../index.js";
import { calculatorFlow } from "../examples/calculator/flow.js";
import { CalculatorService } from "../examples/calculator/rpc-service.js";

const args = new Set(process.argv.slice(2));
const profile = ((process.argv[process.argv.indexOf("--profile") + 1] as RuntimeConfig["profile"]) || process.env.NEON_PROFILE || "auto") as RuntimeConfig["profile"];
const runId = randomUUID();
let sequence = 0;
const emit = (stage: string, data: Record<string, unknown>) => console.log(JSON.stringify({ run_id: runId, stage, sequence: sequence++, ...data }));

function diagnose(producer: Record<string, any>, consumer: Record<string, any>, expected?: Record<string, any>): string {
  if (!consumer.event_id) return "missing_data";
  if ((consumer.input_revision ?? 0) < (producer.input_revision ?? 0)) return "stale_data";
  if (expected && (producer.source_key !== expected.source_key || consumer.target_key !== expected.target_key)) return "coordinate_mismatch";
  if (expected?.actual_revision > expected?.expected_revision && expected?.marked_stale) return "comparison_direction_error";
  return "matched";
}

function diagnostics(): void {
  const cases: Array<[Record<string, any>, Record<string, any>, Record<string, any> | undefined]> = [
    [{ event_id: "evt-missing", input_revision: 1 }, {}, undefined],
    [{ event_id: "evt-stale", input_revision: 4 }, { event_id: "evt-stale", input_revision: 3 }, undefined],
    [{ event_id: "evt-coordinate", input_revision: 1, source_key: "backpack.compass" }, { event_id: "evt-coordinate", input_revision: 1, target_key: "wrong-zone" }, { source_key: "backpack.compass", target_key: "equipment-zone" }],
    [{ event_id: "evt-direction", input_revision: 1 }, { event_id: "evt-direction", input_revision: 2 }, { expected_revision: 1, actual_revision: 2, marked_stale: true }],
    [{ event_id: "evt-match", input_revision: 1, source_key: "backpack.compass" }, { event_id: "evt-match", input_revision: 1, target_key: "equipment-zone" }, { source_key: "backpack.compass", target_key: "equipment-zone" }],
  ];
  for (const [producer, consumer, expected] of cases) emit("diagnostic.case", { input: { producer, consumer }, producer, consumer, result: diagnose(producer, consumer, expected), pass_result: true });
}

const runtime = new RuntimeSession({ ...(process.env.NEON_ROOT ? { neonRoot: process.env.NEON_ROOT } : {}), profile, mode: "headless", domain: "127.0.0.1:39104" });
const service = new CalculatorService();
try {
  if (args.has("--diagnostic")) diagnostics();
  await service.start();
  await runtime.start();
  const ui = new UiClient(new NeonClient(runtime.config.ui, { origin: "inventory-sdk-probe", kind: "cli" }));
  const renderer = new NeonClient(runtime.config.wgpu, { origin: "inventory-sdk-probe", kind: "cli" });
  const program = await ui.submitFlow(await calculatorFlow());
  emit("flow.produced", { input: { flow: "calculator.nui", vertical_slice: "inventory" }, producer: { surface_id: program.surface_id, program_revision: program.program_revision, renderer_epoch: (await renderer.describe("wgpu-runtime")).epoch }, result: "passed", pass_result: true });
  const eventId = randomUUID();
  const event: any = { event_id: eventId, kind: "activate", intent: "calculator.number.one", source_node_key: "one", payload: {}, program_revision: program.program_revision, input_revision: 0, request_id: eventId, idempotency_key: `intent:${eventId}`, interaction: { interaction_id: eventId, sequence: 1, renderer_epoch: 1 } };
  const response = await ui.hostInbound({ kind: "semantic_intent", event });
  const fragmentResponse = await renderer.call("wgpu-runtime", "wgpu.ui.fragment.snapshot", { fragment_id: program.surface_id });
  const fragment: any = fragmentResponse.result ?? {};
  const producer = { event_id: eventId, input_revision: 0, renderer_epoch: 1, source_key: "one" };
  const consumer = { event_id: response ? eventId : null, input_revision: 1, fragment_revision: fragment.fragment_revision ?? null, frame_sequence: fragment.sequence ?? null };
  const diagnostic = diagnose(producer, consumer);
  emit("frame.consumed", { input: { intent: event.intent }, producer, consumer, pairing: { event_id: eventId, status: diagnostic }, result: diagnostic === "matched" ? "passed" : "failed", pass_result: diagnostic === "matched" });
  const passed = diagnostic === "matched" && service.domain.state.display === 1;
  emit("result", { input: { event_count: 1 }, producer: { event_id: eventId, input_revision: 0 }, consumer: { input_revision: 1, fragment_revision: fragment.fragment_revision ?? null }, result: passed ? "passed" : "failed", diagnostic, pass_result: passed });
  process.exitCode = passed ? 0 : 1;
} catch (error) {
  emit("result", { result: "failed", diagnostic: "missing_data", pass_result: false, error: String(error) });
  process.exitCode = 1;
} finally {
  await runtime.stop();
  await service.stop();
}
